### ISSUE TITLE
Implementation of `Debug` and `Hash` is not general enough

### ISSUE THREAD

**Issue by nlinker** [kotauskas/thin_trait_object#1]
Hi. 
The crate looks promising, however I bumped with a pretty cryptic error. MRE:
```rust
use thin_trait_object::*;

#[derive(Debug, Clone)]
struct Bar(String);

#[thin_trait_object]
trait Foo {
    fn fooify(&self, bar: &Bar);
}
impl Foo for String {
    fn fooify(&self, bar: &Bar) {
        println!("Fooified a string: {} {:?}", self, bar);
    }
}

fn main() {
    let bar = Bar("suffix".to_string());
    BoxedFoo::new("Hello World!".to_string()).fooify(&bar);
}
```
The error:
```
➜  thin_trait_object git:(main) ✗ cargo run --example basic
   Compiling thin_trait_object v1.1.1 (/Users/nick/rust/scai/utils/cxx-all/thin_trait_object)
error: implementation of `Hash` is not general enough
 --> examples/basic.rs:6:1
  |
6 | #[thin_trait_object]
  | ^^^^^^^^^^^^^^^^^^^^ implementation of `Hash` is not general enough
  |
  = note: `Hash` would have to be implemented for the type `for<'r> unsafe fn(*mut c_void, &'r Bar)`
  = note: ...but `Hash` is actually implemented for the type `unsafe fn(*mut c_void, &'0 Bar)`, for some specific lifetime `'0`
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: implementation of `Debug` is not general enough
 --> examples/basic.rs:6:1
  |
6 | #[thin_trait_object]
  | ^^^^^^^^^^^^^^^^^^^^ implementation of `Debug` is not general enough
  |
  = note: `Debug` would have to be implemented for the type `for<'r> unsafe fn(*mut c_void, &'r Bar)`
  = note: ...but `Debug` is actually implemented for the type `unsafe fn(*mut c_void, &'0 Bar)`, for some specific lifetime `'0`
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors
```
I also put the MRE into the fork https://github.com/nlinker/thin_trait_object/blob/main/examples/basic.rs

**Comment by kotauskas** [comment]
This is indeed so cryptic that I suspect it may be a compiler bug. What's happening here is that it's trying to compile `FooVtable`'s definition which looks like this:
```rust
#[derive(Copy, Clone, Debug, Hash)]
struct FooVtable {
    pub fooify: unsafe fn(*mut ::core::ffi::c_void, &Bar),
    pub drop: unsafe fn(*mut ::core::ffi::c_void),
}
```
where the `unsafe fn(*mut ::core::ffi::c_void, &Bar)` desugars into `for<'r> unsafe fn(*mut ::core::ffi::c_void, &'r Bar)`. The `#[derive(Debug, Hash)]` derives then try to use the implementation of `Debug` on that type, `for<'r> unsafe fn(*mut ::core::ffi::c_void, &'r Bar)`, which is the type of the field, but that fails for... some reason. It really shouldn't: `Debug` should be implemented for all `fn` types and the implicit HRTB, `for<'r>`, has nothing to do with the function pointer's ability to be printed.

I'm currently writing a band-aid fix to this (implementing `Debug` and `Hash` manually by casting the function pointer to a `*mut ::core::ffi::c_void`), and will explore the compiler bug further if I can.

**Comment by kotauskas** [comment]
This indeed turned out to be a [known compiler bug][70263]. I've just released a fix in 1.1.2 which is functionally identical to how the derives would behave once the compiler bug is fixed.

[70263]: https://github.com/rust-lang/rust/issues/70263#issuecomment-689852550 " "
